File: python/inkbox/phone/resources/webhooks.py
"""
inkbox/phone/resources/webhooks.py

Webhook CRUD for phone numbers.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from uuid import UUID

from inkbox.phone.types import PhoneWebhook, PhoneWebhookCreateResult

if TYPE_CHECKING:
    from inkbox.phone._http import HttpTransport

_UNSET = object()


class PhoneWebhooksResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def create(
        self,
        phone_number_id: UUID | str,
        *,
        url: str,
        event_types: list[str],
    ) -> PhoneWebhookCreateResult:
        """Create a webhook for a phone number.

        Args:
            phone_number_id: UUID of the phone number.
            url: HTTPS URL to receive webhook events.
            event_types: List of event types to subscribe to.

        Returns:
            The created webhook including the signing secret.
        """
        data = self._http.post(
            f"/numbers/{phone_number_id}/webhooks",
            json={"url": url, "event_types": event_types},
        )
        return PhoneWebhookCreateResult._from_dict(data)

    def list(self, phone_number_id: UUID | str) -> list[PhoneWebhook]:
        """List webhooks for a phone number."""
        data = self._http.get(f"/numbers/{phone_number_id}/webhooks")
        return [PhoneWebhook._from_dict(w) for w in data]

    def update(
        self,
        phone_number_id: UUID | str,
        webhook_id: UUID | str,
        *,
        url: str | None = _UNSET,  # type: ignore[assignment]
        event_types: list[str] | None = _UNSET,  # type: ignore[assignment]
    ) -> PhoneWebhook:
        """Update a webhook.

        Pass only the fields you want to change; omitted fields are left as-is.

        Args:
            phone_number_id: UUID of the phone number.
            webhook_id: UUID of the webhook.
            url: New HTTPS URL for the webhook.
            event_types: New list of event types.
        """
        body: dict[str, Any] = {}
        if url is not _UNSET:
            body["url"] = url
        if event_types is not _UNSET:
            body["event_types"] = event_types
        data = self._http.patch(
            f"/numbers/{phone_number_id}/webhooks/{webhook_id}",
            json=body,
        )
        return PhoneWebhook._from_dict(data)

    def delete(self, phone_number_id: UUID | str, webhook_id: UUID | str) -> None:
        """Delete a webhook.

        Args:
            phone_number_id: UUID of the phone number.
            webhook_id: UUID of the webhook to delete.
        """
        self._http.delete(f"/numbers/{phone_number_id}/webhooks/{webhook_id}")

File: typescript/src/phone/resources/webhooks.ts
/**
 * inkbox-phone/resources/webhooks.ts
 *
 * Webhook CRUD for phone numbers.
 */

import { HttpTransport } from "../../_http.js";
import {
  PhoneWebhook,
  PhoneWebhookCreateResult,
  RawPhoneWebhook,
  RawPhoneWebhookCreateResult,
  parsePhoneWebhook,
  parsePhoneWebhookCreateResult,
} from "../types.js";

const BASE = "/numbers";

export class PhoneWebhooksResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Create a webhook subscription for a phone number.
   *
   * @param phoneNumberId - UUID of the phone number.
   * @param options.url - HTTPS URL to receive webhook events.
   * @param options.eventTypes - List of event types to subscribe to.
   */
  async create(
    phoneNumberId: string,
    options: { url: string; eventTypes: string[] },
  ): Promise<PhoneWebhookCreateResult> {
    const data = await this.http.post<RawPhoneWebhookCreateResult>(
      `${BASE}/${phoneNumberId}/webhooks`,
      { url: options.url, event_types: options.eventTypes },
    );
    return parsePhoneWebhookCreateResult(data);
  }

  /** List all webhooks for a phone number. */
  async list(phoneNumberId: string): Promise<PhoneWebhook[]> {
    const data = await this.http.get<RawPhoneWebhook[]>(
      `${BASE}/${phoneNumberId}/webhooks`,
    );
    return data.map(parsePhoneWebhook);
  }

  /**
   * Update a webhook subscription. Only provided fields are updated.
   *
   * @param phoneNumberId - UUID of the phone number.
   * @param webhookId - UUID of the webhook.
   * @param options.url - New HTTPS URL.
   * @param options.eventTypes - New list of event types.
   */
  async update(
    phoneNumberId: string,
    webhookId: string,
    options: { url?: string; eventTypes?: string[] },
  ): Promise<PhoneWebhook> {
    const body: Record<string, unknown> = {};
    if (options.url !== undefined) {
      body["url"] = options.url;
    }
    if (options.eventTypes !== undefined) {
      body["event_types"] = options.eventTypes;
    }
    const data = await this.http.patch<RawPhoneWebhook>(
      `${BASE}/${phoneNumberId}/webhooks/${webhookId}`,
      body,
    );
    return parsePhoneWebhook(data);
  }

  /** Delete a webhook subscription. */
  async delete(phoneNumberId: string, webhookId: string): Promise<void> {
    await this.http.delete(
      `${BASE}/${phoneNumberId}/webhooks/${webhookId}`,
    );
  }
}
